sopel/irc/__init__.py

```diff
@@ -584,15 +592,28 @@ def say(self, text, recipient, max_messages=1):
             # based on how long it's been since our last message to recipient
             if not recipient_stack['flood_left']:
                 recipient_stack['flood_left'] = min(
-                    self.config.core.flood_burst_lines,
-                    int(elapsed) * self.config.core.flood_refill_rate)
+                    flood_burst_lines,
+                    int(elapsed) * flood_refill_rate)
 
             # If it's too soon to send another message, wait
             if not recipient_stack['flood_left']:
-                penalty = float(max(0, len(text) - 50)) / 70
-                wait = min(self.config.core.flood_empty_wait + penalty, 2)  # Maximum wait time is 2 sec
+                penalty = 0
+
+                if flood_penalty_ratio > 0:
+                    penalty_ratio = flood_text_length * flood_penalty_ratio
+                    text_length_overflow = float(
+                        max(0, len(text) - flood_text_length))
+                    penalty = text_length_overflow / penalty_ratio
+
+                # Maximum wait time is 2 sec
+                wait = min(flood_empty_wait + penalty, flood_max_wait)
                 if elapsed < wait:
-                    time.sleep(wait - elapsed)
+                    sleep_time = wait - elapsed
+                    LOGGER.debug(
+                        'Flood protection: waiting %.3fs '
+                        '(initial wait of %.3fs with penalty of %.3fs, max 2s)',
+                        sleep_time, wait, penalty)
```

Comment:
```suggestion
                        '(initial wait of %.3fs with penalty of %.3fs; max %.3fs)',
                        sleep_time, wait, penalty, flood_max_wait)
```

I think `elapsed` needs to be factored in here too, or it'll be confusing to see some `sleep_time` values that don't add up to `wait + penalty` in logs. Not sure how to work it into the message rn though; I'm sleepy. 😪

Reply:
Will do.